File: app/src/main/java/ru/geekbrains/githubclient/mvp/presenter/UsersPresenter.java
package ru.geekbrains.githubclient.mvp.presenter;

import android.util.Log;

import androidx.appcompat.view.menu.MenuView;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import moxy.MvpPresenter;
import ru.geekbrains.githubclient.GithubApplication;
//import ru.geekbrains.githubclient.LoginActivity;
import ru.geekbrains.githubclient.MainActivity;
import ru.geekbrains.githubclient.mvp.model.entity.GithubUser;
import ru.geekbrains.githubclient.mvp.model.entity.GithubUserRepo;
import ru.geekbrains.githubclient.mvp.presenter.list.IUserListPresenter;
import ru.geekbrains.githubclient.mvp.view.IItemView;
import ru.geekbrains.githubclient.mvp.view.UserItemView;
import ru.geekbrains.githubclient.mvp.view.UsersView;
import ru.geekbrains.githubclient.navigation.Screens;
import ru.terrakok.cicerone.Router;

public class UsersPresenter extends MvpPresenter<UsersView> {
    private static final String TAG = UsersPresenter.class.getSimpleName();

    private static final boolean VERBOSE = true;

    private GithubUserRepo usersRepo = new GithubUserRepo();
    private Router router = GithubApplication.getApplication().getRouter();

    private class UsersListPresenter implements IUserListPresenter {

        private List<GithubUser> users = new ArrayList<>();

        @Override
        public void onItemClick(UserItemView view) {
            if (VERBOSE) {
                Log.v(TAG, " onItemClick " + view.getPos());

                router.navigateTo(new Screens.LoginScreen(view.getPos()));

            }

        }

        @Override
        public void bindView(UserItemView view) {
            GithubUser user = users.get(view.getPos());
            view.setLogin(user.getLogin());
        }

        @Override
        public int getCount() {
            return users.size();
        }
    }

    private UsersListPresenter usersListPresenter = new UsersListPresenter();

    public UsersListPresenter getUsersListPresenter() {
        return usersListPresenter;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        getViewState().init();
        loadData();

    }


    private void loadData() {
        List<GithubUser> users = usersRepo.getUsers();
        usersListPresenter.users.addAll(users);
        getViewState().updateList();
    }

}

File: app/src/main/java/ru/geekbrains/githubclient/mvp/presenters/UsersPresenter.kt
package ru.geekbrains.githubclient.mvp.presenters

import android.util.Log
import io.reactivex.rxjava3.core.Scheduler

import moxy.MvpPresenter
import ru.geekbrains.githubclient.GithubApplication
import ru.geekbrains.githubclient.mvp.model.entity.GithubUser
import ru.geekbrains.githubclient.mvp.model.repo.IGithubUsersRepo
import ru.geekbrains.githubclient.mvp.presenters.list.ILoginListPresenter
import ru.geekbrains.githubclient.mvp.view.list.LoginItemView
import ru.geekbrains.githubclient.mvp.view.UsersView
import ru.geekbrains.githubclient.navigation.Screens
import ru.terrakok.cicerone.Router
import javax.inject.Inject

class UsersPresenter : MvpPresenter<UsersView>() {
    @Inject
    lateinit var scheduler: Scheduler
    @Inject
    lateinit var usersRepo: IGithubUsersRepo
    @Inject
    lateinit var router: Router
    init {
        GithubApplication.instance.appComponent.inject(this)
    }


    class UsersListPresenter : ILoginListPresenter {

        val users = mutableListOf<GithubUser>()

        override var itemClickListener: ((LoginItemView) -> Unit)? = null

        override fun getCount() = users.size

        override fun bindView(view: LoginItemView) {
            val user = users[view.pos]
            user.login?.let { view.setLogin(it)}
            user.avatarUrl?.let { view.loadAvatar(it) }
        }

    }

    val usersListPresenter = UsersListPresenter()

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.init()
        loadData()

        usersListPresenter.itemClickListener = {
            val user: GithubUser = usersListPresenter.users[it.pos]
            router.navigateTo(Screens.LoginScreen(user))
        }
    }

    private fun loadData() {
        usersRepo.getUsers()
                .observeOn(scheduler)
                .subscribe({ repos ->
                    usersListPresenter.users.clear()
                    usersListPresenter.users.addAll(repos)
                    viewState.updateList()
                }, { error -> (Log.e("log", "Error: ${error}")) })
    }
    fun backPressed(): Boolean {
        router.exit()
        return true
    }
}



